src/mlpack/methods/ann/layer/residual_impl.hpp

```diff
@@ -0,0 +1,209 @@
+ * @author Shubham Agrawal
+ *
+ * Implementation of the base class for neural network layers that are wrappers
+ * around other layers.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_RESIDUAL_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_RESIDUAL_IMPL_HPP
+
+#include "residual.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<typename MatType>
+ResidualType<MatType>::ResidualType() :
+    MultiLayer<MatType>()
+{
+  // Nothing to do.
+}
+
+template<typename MatType>
+ResidualType<MatType>::ResidualType(const ResidualType& other) :
+    MultiLayer<MatType>(other)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+ResidualType<MatType>::ResidualType(ResidualType&& other) :
+    MultiLayer<MatType>(other)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+ResidualType<MatType>& ResidualType<MatType>::operator=(const ResidualType& other)
+{
+  if (this != &other)
+  {
+    MultiLayer<MatType>::operator=(other);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+ResidualType<MatType>& ResidualType<MatType>::operator=(ResidualType&& other)
+{
+  if (this != &other)
+  {
+    MultiLayer<MatType>::operator=(other);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void ResidualType<MatType>::Forward(
+    const MatType& input, MatType& output)
+{
+  // Make sure training/testing mode is set right in each layer.
+  for (size_t i = 0; i < this->network.size(); ++i)
+    this->network[i]->Training() = this->training;
+
+  // Note that we use `output` for the last layer; layerOutputs is only used for
+  // intermediate values between layers.
+  if (this->network.size() > 1)
+  {
+    // Initialize memory for the forward pass (if needed).
+    this->InitializeForwardPassMemory(input.n_cols);
+
+    for (size_t i = 0; i < this->network.size(); i++)
+      this->network[i]->Forward(input, this->layerOutputs[i]);
+
+		// Reduce the outputs to single output.
+    output.zeros();
+		for (size_t i = 0; i < this->layerOutputs.size(); i++)
+		{
+			output += this->layerOutputs[i];
+		}
+  }
+  else if (this->network.size() == 1)
+  {
+    this->network[0]->Forward(input, output);
+  }
+  else
+  {
+    // Empty network?
+    output = input;
+  }
+}
+
+template<typename MatType>
+void ResidualType<MatType>::Backward(
+    const MatType& input, const MatType& gy, MatType& g)
+{
+  if (this->network.size() > 1)
+  {
+    // Initialize memory for the backward pass (if needed).
+    this->InitializeBackwardPassMemory(input.n_cols);
+
+    g.zeros();
+    for (size_t i = 0; i < this->network.size(); i++) {
+      this->network[i]->Backward(this->layerOutputs[i], gy, this->layerDeltas[i]);
+			g += this->layerDeltas[i];
+		}
```

Comment:
Looks like the style is off here too.  Opening braces should be on a new line, for instance.